File: 208-implement-trie-prefix-tree/208-implement-trie-prefix-tree.cs
public class Trie {
    class TrieNode
    {
        public Dictionary<char, TrieNode> edges = new Dictionary<char, TrieNode>();
        public bool isWordEnd;
        
        public TrieNode()
        {
        }
    }
    
    private TrieNode root = new TrieNode();
    public Trie()
    {
    }
    
    public void Insert(string word) {
        TrieNode currentNode = root;
        foreach (char c in word)
        {
            if (! currentNode.edges.ContainsKey(c))
            {
                currentNode.edges[c] = new TrieNode();
            }
            currentNode = currentNode.edges[c];
        }
        currentNode.isWordEnd = true;
    }
    
    public bool Search(string word)
    {
        TrieNode currentNode = root;
        
        foreach (char c in word)
        {
            if (! currentNode.edges.ContainsKey(c))
            {
                return false;
            }
            currentNode = currentNode.edges[c];
        }
        
        return currentNode.isWordEnd;
        
    }
    
    public bool StartsWith(string prefix)
    {
        TrieNode currentNode = root;
        
        foreach (char c in prefix)
        {
            if (! currentNode.edges.ContainsKey(c))
            {
                return false;
            }
            currentNode = currentNode.edges[c];
        }
        
        return true;
        
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.Insert(word);
 * bool param_2 = obj.Search(word);
 * bool param_3 = obj.StartsWith(prefix);
 */

File: 208-implement-trie-prefix-tree/208-implement-trie-prefix-tree.java
class Trie {
    class Node {
        boolean isWord;
        Map<Character, Node> children = new HashMap();
    }

    Node root;
    
    public Trie() {
        this.root = new Node();
    }
    
    public void insert(String word) {
        var curr = root;
        
        for (var c : word.toCharArray()) {
            if (!curr.children.containsKey(c)) {
                curr.children.put(c, new Node());
            }
            curr = curr.children.get(c);
        }
        
        curr.isWord = true;
    }
    
    public boolean search(String word) {
        var curr = root;
        
        for (var c : word.toCharArray()) {
            if (!curr.children.containsKey(c)) {
                return false;
            }
            curr = curr.children.get(c);
        }
        
        return curr.isWord;
    }
    
    public boolean startsWith(String prefix) {
        var curr = root;
        
        for (var c : prefix.toCharArray()) {
            if (!curr.children.containsKey(c)) {
                return false;
            }
            curr = curr.children.get(c);
        }
        
        return true;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */